### PR TITLE
[#772] txQueueManager to not depend on AccountManager and nodeManager

### LINKED ISSUE #772: Transaction manager cleanup
# Problem

TxQueueManager has a dependency on `Accounter` interface and `NodeManager`

# Acceptance Criteria

- [x] "TxQueueManager" does not depend on `Accounter` interface
- [x] "TxQueueManager" does not depend on nodeManager
- [x] "Accounter" interface is removed
- [x] "Accounter" mocks are removed

# Future Steps

- Remove common.NodeManager interface and its mocks
- Move to transactions specific logic from common to transactions package

### PR DESCRIPTION
**Pull request by adriacidre**
Changes:

* [x] Removed TxQueueManager dependency on AccountManager by moving account validation logic to StatusBackend
* [x] Removed Accounter interface
* [x] Removed Accounter Mocks
* [x] Added new tests on transactions package for missing methods
* [x] Remove TxQueueManager dependency on nodeManager by 
    * [x] introducing RPCClientProvider same as it’s done on JailManager
    * [x] Injecting NetworkID on Start method

Next steps:

* ​We should be fine to remove common.NodeManager interface and mocks
* Move to transactions specific logic from common to transactions package

Closes #772

### CODE REVIEW

File: geth/api/backend.go [hunk 1]
```diff
@@ -208,14 +208,46 @@ func (b *StatusBackend) SendTransaction(ctx context.Context, args common.SendTxA
...
+		b.log.Warn("failed to verify account", "account", selectedAccount.Address.String(), "error", err)
+		return selectedAccount, err
+	}
+	return selectedAccount, nil
+}
+
 // CompleteTransaction instructs backend to complete sending of a given transaction
 func (b *StatusBackend) CompleteTransaction(id common.QueuedTxID, password string) (gethcommon.Hash, error) {
-	return b.txQueueManager.CompleteTransaction(id, password)
+	selectedAccount, verificationErr := b.getVerifiedAccount(password)
```
Comment (dshulyak) [inline geth/api/backend.go]: why you can't exit here if the account is nil or there was a verification err?

**Review by dshulyak — CHANGES_REQUESTED** [review]
nice changes, can you explain what is the reason to pass verification error down the stack?

File: geth/api/backend.go (hunk 1 above)
Comment (adriacidre) [reply geth/api/backend.go]: Sure, I know this can result a bit ugly :(. 

Basically `txQueuemanager` [removes transaction from queue if no error, or error is not transient](https://github.com/status-im/status-go/pull/773/files/ffdbab782fc3a62a6b74766f966b11025993a232#diff-a16061196c3bf37b6badd79a450b352aL146). 

To keep the changes compatible with current behaviour txQueueManager needs to have access to the error generated by the account verification.

Comment (adriacidre) [reply geth/api/backend.go]: Sure, I know this can result a bit ugly :(

Basically `txQueuemanager.CompleteTransaction` [removes transaction from queue if no error, or error is not transient](https://github.com/status-im/status-go/pull/773/files/ffdbab782fc3a62a6b74766f966b11025993a232#diff-a16061196c3bf37b6badd79a450b352aL146). 

To keep the changes compatible with current behaviour txQueueManager needs to have access to the error generated by the account verification.

[This test](https://github.com/status-im/status-go/blob/ffdbab782fc3a62a6b74766f966b11025993a232/t/e2e/transactions/transactions_test.go#L415) raised the problem in first instance.

Comment (themue) [reply geth/api/backend.go]: Could you add a little comment about it for a later maintainer?

Comment (dshulyak) [reply geth/api/backend.go]: right, so 2/3 transient errors can be avoided if we will exit here with an error or if the account is nil (1 and 3 in the list below). but there is still a notification... please check if we can easily move NotifyOnReturn up the stack? If not then fine

```
keystore.ErrDecrypt.Error():          true, // wrong password
ErrInvalidCompleteTxSender.Error():   true, // completing tx create from another account
account.ErrNoAccountSelected.Error(): true, // account not selected
```

Comment (adriacidre) [reply geth/api/backend.go]: cheers, i'll have a look

File: geth/api/backend.go [hunk 2]
```diff
@@ -208,14 +208,46 @@ func (b *StatusBackend) SendTransaction(ctx context.Context, args common.SendTxA
 	return rst.Hash, nil
 }
 
+func (b *StatusBackend) getVerifiedAccount(password string) (*account.SelectedExtKey, error) {
+	selectedAccount, err := b.accountManager.SelectedAccount()
+	if err != nil {
+		b.log.Warn("failed to get a selected account", "err", err)
+		return nil, err
+	}
+	// make sure that only account which created the tx can complete it
```
Comment (adambabik) [inline geth/api/backend.go]: This method is not related to transactions so this commit seems invalid.

File: geth/api/backend.go [hunk 3]
```diff
@@ -208,14 +208,46 @@ func (b *StatusBackend) SendTransaction(ctx context.Context, args common.SendTxA
 	return rst.Hash, nil
 }
 
+func (b *StatusBackend) getVerifiedAccount(password string) (*account.SelectedExtKey, error) {
+	selectedAccount, err := b.accountManager.SelectedAccount()
+	if err != nil {
+		b.log.Warn("failed to get a selected account", "err", err)
```
Comment (adambabik) [inline geth/api/backend.go]: If we return an error then maybe we should also print it as an error?

File: geth/api/backend.go [hunk 4]
```diff
@@ -208,14 +208,46 @@ func (b *StatusBackend) SendTransaction(ctx context.Context, args common.SendTxA
...
+		return nil, err
+	}
+	// make sure that only account which created the tx can complete it
+	config, err := b.NodeManager().NodeConfig()
+	if err != nil {
+		return selectedAccount, err
```
Comment (adambabik) [inline geth/api/backend.go]: I am not sure we should return selected account if the verification was not passed.

**Review by adambabik — COMMENTED** [review]
Great stuff! Let's see how it's gonna work if we move signals and transient errors to `StatusBackend`.

File: geth/api/backend.go (hunk 1 above)
Comment (adriacidre) [reply geth/api/backend.go]: I've exposed `TxQueueManager.NotifyErrored` [on this commit](https://github.com/status-im/status-go/pull/773/commits/82510a2cf89cdd84417794713ecc8199a228f2c6) and linked it from StatusBackend.

Not sure it makes sense use it that way, could you please have a look?

File: geth/transactions/txqueue_manager.go [hunk 5]
```diff
@@ -125,8 +131,23 @@ func (m *Manager) WaitForTransaction(tx *common.QueuedTx) common.TransactionResu
 	}
 }
 
+// NotifyErrored sends a notification for the given transaction
+func (m *Manager) NotifyErrored(id common.QueuedTxID, inputError error) error {
```
Comment (dshulyak) [inline geth/transactions/txqueue_manager.go:135]: it would be better to move notifications to backend if possible, but this is also fine

**Review by dshulyak — APPROVED** [review]

File: geth/transactions/txqueue_manager.go (hunk 5 above)
Comment (adriacidre) [reply geth/transactions/txqueue_manager.go:135]: Ok, I'll add a new point to https://github.com/status-im/ideas/issues/63 to extract notifications to backend.

File: geth/api/backend.go [hunk 6]
```diff
@@ -208,14 +208,47 @@ func (b *StatusBackend) SendTransaction(ctx context.Context, args common.SendTxA
...
+		return nil, err
+	}
+	return selectedAccount, nil
+}
+
 // CompleteTransaction instructs backend to complete sending of a given transaction
-func (b *StatusBackend) CompleteTransaction(id common.QueuedTxID, password string) (gethcommon.Hash, error) {
-	return b.txQueueManager.CompleteTransaction(id, password)
+func (b *StatusBackend) CompleteTransaction(id common.QueuedTxID, password string) (hash gethcommon.Hash, err error) {
+	selectedAccount, err := b.getVerifiedAccount(password)
```
Comment (dshulyak) [inline geth/api/backend.go:231]: one more thing, could it be that error is nil and the selected account is nil too? there was a distinct error for such case before, but might be just a legacy

Comment (adriacidre) [reply geth/api/backend.go:231]: `getVerifierAccount` is using [account.SelectedAccount](https://github.com/status-im/status-go/blob/498924f5be57dfea825534873411ca671ba557ad/geth/account/accounts.go#L239) which is returning an error if selectedAccount is nil. So, in the current approach should not possible to have such situation.

However, anybody could directly call tx.CompleteAccount with nil account which will produce a panic [here](https://github.com/status-im/status-go/blob/498924f5be57dfea825534873411ca671ba557ad/geth/transactions/txqueue_manager.go#L175) I'll add a defensive check there and return an `account.ErrNoAccountSelected`. That should keep the same behaviour as before :)

```
	if selectedAccount == nil {
		return account.ErrNoAccountSelected
	}
```

File: geth/api/backend.go [hunk 7]
```diff
@@ -208,14 +208,47 @@ func (b *StatusBackend) SendTransaction(ctx context.Context, args common.SendTxA
...
+	if err != nil {
+		_ = b.txQueueManager.NotifyErrored(id, err)
+		return hash, err
+	}
+
+	hash, err = b.txQueueManager.CompleteTransaction(id, selectedAccount)
```
Comment (adambabik) [inline geth/api/backend.go]: Nitpick: does `return b.txQueueManager.CompleteTransaction(id, selectedAccount)` work?

**Review by adambabik — APPROVED** [review]
LGTM 🎉

File: geth/api/backend.go (hunk 7 above)
Comment (adriacidre) [reply geth/api/backend.go]: mmm, sure, should work fine, i'll change it!

File: geth/api/backend.go [hunk 8]
```diff
@@ -208,14 +208,46 @@ func (b *StatusBackend) SendTransaction(ctx context.Context, args common.SendTxA
...
+		return hash, err
+	}
+
+	return b.txQueueManager.CompleteTransaction(id, selectedAccount)
 }
 
 // CompleteTransactions instructs backend to complete sending of multiple transactions
 func (b *StatusBackend) CompleteTransactions(ids []common.QueuedTxID, password string) map[common.QueuedTxID]common.TransactionResult {
-	return b.txQueueManager.CompleteTransactions(ids, password)
+	results := make(map[common.QueuedTxID]common.TransactionResult)
```
Comment (mandrigin) [inline geth/api/backend.go:242]: That isn't necessary, but if you want to store only a result *either* an error you can use interfaces like that: https://play.golang.org/p/-ANFxqKwPsM

**Review by mandrigin — APPROVED** [review]
👍

File: geth/api/backend.go (hunk 8 above)
Comment (adriacidre) [reply geth/api/backend.go:242]: Nice, i always forget about type switches. I guess i have the preconceived  idea they don't have a good performance :/
I'll try to use this approach next time :)
